dll/win32/shell32/shlexec.cpp

```diff
@@ -458,6 +458,132 @@ static HRESULT SHELL_GetPathFromIDListForExecuteW(LPCITEMIDLIST pidl, LPWSTR psz
     return hr;
 }
 
+static HWND SHELL_GetUsableDialogOwner(HWND hwnd)
+{
+    // Explicitly block the shell desktop listview from becoming the owner (IContextMenu calling ShellExecute)
+    HWND progman = GetShellWindow();
+    if (hwnd && IsWindowVisible(hwnd) && hwnd != GetDesktopWindow() &&
+        hwnd != progman && !IsChild(progman, hwnd))
+    {
+        return hwnd;
+    }
+    return NULL;
+}
+
+/*************************************************************************
+ *    SHELL_PromptAndRunProcessAs [Internal]
+ */
+typedef struct {
+    WCHAR NameBuffer[MAX_PATH], Pass[MAX_PATH];
+    LPWSTR Name, Domain;
+    BOOL Safer;
+} RUNASDLGDATA;
+
+static INT_PTR CALLBACK RunAsDlgProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    RUNASDLGDATA *pData = (RUNASDLGDATA *)GetWindowLongPtr(hwnd, DWLP_USER);
+    switch(uMsg)
+    {
+        case WM_INITDIALOG:
+            {
+                SetWindowLongPtr(hwnd, DWLP_USER, (LPARAM)(pData = (RUNASDLGDATA *)lParam));
+                SendDlgItemMessage(hwnd, IDC_RUNAS_NAME, EM_LIMITTEXT, _countof(RUNASDLGDATA::NameBuffer)-1, 0);
+                SendDlgItemMessage(hwnd, IDC_RUNAS_PASS, EM_LIMITTEXT, _countof(RUNASDLGDATA::Pass)-1, 0);
+                SendDlgItemMessage(hwnd, IDC_RUNAS_OTHER, BM_CLICK, 0, 0);
+
+                HWND control = GetDlgItem(hwnd, IDC_RUNAS_THIS);
+                WCHAR fmtbuf[200], buf[_countof(fmtbuf) + _countof(RUNASDLGDATA::NameBuffer)];
+                DWORD cch = _countof(RUNASDLGDATA::NameBuffer);
+                if (GetUserName(pData->NameBuffer, &cch))
```

Comment:
Since this is all new code, please:
1. use explicit unicode function names;
2. use string-safe functions (those wine-ish wsprintfW are deprecated).